p2p/net/conn_test.go

```diff
@@ -48,8 +48,57 @@ func TestSendMessage(t *testing.T) {
 	case <-rwcam.writeWaitChan:
 		rcvmsg := <-rwcam.writeWaitChan
 		assert.Equal(t, rcvmsg, []byte(msg))
-	case <-time.After(5 * time.Second):
-		assert.Fail(t, "timeout waiting for message to be sent")
+	case <-time.After(2 * time.Second):
+		assert.Fail(t, "timed out waiting for message to be sent")
+	}
+}
+
+func TestSendMutex(t *testing.T) {
+	netw := NewNetworkMock()
+	rwcam := NewReadWriteCloseAddresserMock()
+	rPub := p2pcrypto.NewRandomPubkey()
+	rwcam.writeWaitChan = make(chan []byte) // unbuffered channel, write is blocking
+	conn := newConnection(rwcam, netw, rPub, &networkSessionImpl{}, msgSizeLimit, time.Second, netw.logger)
+	go conn.beginEventProcessing(context.TODO())
+	msg := "hello"
+	err := conn.Send(context.TODO(), []byte(msg))
+	assert.NoError(t, err)
+
+	// Make sure SendSock received the message and tried to send it (and is blocking on Write)
+	for i := 0; i < 10; i++ {
+		if rwcam.WriteCount() >= 1 {
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+	if rwcam.WriteCount() < 1 {
+		assert.Fail(t, "timed out waiting for SendSock to block")
+	}
+
+	// Now try sending another message and make sure it doesn't block
+	done := make(chan struct{})
+	go func() {
+		err := conn.Send(context.TODO(), []byte(msg))
+		assert.NoError(t, err)
+		close(done)
+	}()
+	select {
+	case <-done:
+		// success
+		break
+	case <-time.After(time.Second):
+		assert.Fail(t, "timed out waiting for message to be sent")
+	}
+
+	// make sure we can read both messages
+	for i := 0; i < 2; i++ {
+		select {
+		case <-rwcam.writeWaitChan:
+			rcvmsg := <-rwcam.writeWaitChan
+			assert.Equal(t, rcvmsg, []byte(msg))
+		case <-time.After(2 * time.Second):
```

Comment:
this is not accurate, check this example - 
https://play.golang.org/p/7o3gQMl-kMi

Reply:
right, I don't think that would work here since we need to read from the channel twice in succession, and if we don't block the second time, we'll miss the second message and fail prematurely

Reply:
> this is not accurate, check this example -
https://play.golang.org/p/7o3gQMl-kMi

The non-`default` case doesn't return (for 2 sec), so `default` is selected. What's not accurate?

> right, I don't think that would work here since we need to read from the channel twice in succession, and if we don't block the second time, we'll miss the second message and fail prematurely

You are entering the `for` loop after the second `Send` returned, so the channel is expected to already have the 2 batches waiting in the channel, and `default` to not be selected. There's still some place for discrepancies due to Go's non-deterministic scheduler, but I usually favor trying to eliminate magic-duration waits.

Reply:
@moshababo you're right, there's no concurrency issue here, the message sends are already synchronized and they should be immediately available. will clean up.

Reply:
Actually, there is a concurrency issue here after all. `Send` is called synchronously, but under the hood the sending mechanism is concurrent, so we need the timeout after all. See [this failure](https://github.com/spacemeshos/go-spacemesh/runs/2796067538?check_suite_focus=true).